include/pybind11/detail/internals.h

```diff
@@ -281,20 +282,59 @@ inline internals **&get_internals_pp() {
 }
 
 inline void translate_exception(std::exception_ptr p) {
+    if (!p) {
+        return;
+    }
+    auto raise_err = PyErr_SetString;
+#if PY_VERSION_HEX >= 0x03030000
+    // handles nested C++ exceptions if supported
     try {
-        if (p) std::rethrow_exception(p);
-    } catch (error_already_set &e)           { e.restore();                                    return;
-    } catch (const builtin_exception &e)     { e.set_error();                                  return;
-    } catch (const std::bad_alloc &e)        { PyErr_SetString(PyExc_MemoryError,   e.what()); return;
-    } catch (const std::domain_error &e)     { PyErr_SetString(PyExc_ValueError,    e.what()); return;
-    } catch (const std::invalid_argument &e) { PyErr_SetString(PyExc_ValueError,    e.what()); return;
-    } catch (const std::length_error &e)     { PyErr_SetString(PyExc_ValueError,    e.what()); return;
-    } catch (const std::out_of_range &e)     { PyErr_SetString(PyExc_IndexError,    e.what()); return;
-    } catch (const std::range_error &e)      { PyErr_SetString(PyExc_ValueError,    e.what()); return;
-    } catch (const std::overflow_error &e)   { PyErr_SetString(PyExc_OverflowError, e.what()); return;
-    } catch (const std::exception &e)        { PyErr_SetString(PyExc_RuntimeError,  e.what()); return;
+        // Rethrow the exception_ptr to recove type info.
+        // Dereferencing an exception_ptr is UB otherwise.
+        std::rethrow_exception(p);
+    } catch (const std::nested_exception &nep) {
+        std::exception_ptr nested = nep.nested_ptr();
+        if (nested != nullptr && nested != p) {
+            translate_exception(nested);
+            raise_err = raise_from;
+        }
+    } catch (...) {
+    }
+#endif
+    try {
+        std::rethrow_exception(p);
+    } catch (error_already_set &e) {
+        e.restore();
```

Comment:
Here and for `builtin_exception` below `raise_err` isn't referenced.
Could these two `catch` be moved up?
Hm, it's getting a bit tricky with the #ifdef unfortunately.
We're almost removing 2.7 and 3.5 support, which would make that idea straightforward.
If you think my suggestion has merit, could you please add a comment, e.g.

// TODO: Move `error_already_set` and `builtin_exception` `catch` blocks up after removal of Python 2 & 3.5 support (because `raise_err` is not needed).

Reply:
I suppose they could be moved up. The only purpose of the first try catch is to figure out if an exception_ptr refers to a nested_exception and recur down on the stack for it. The two try blocks could be merged if I duplicated the catch logic for the handling of the nested_exception.

(Alternatively, I could write a function that handles a nested exception on every single catch block. That would remove the try catch, but complicate other parts like raise-err selection etc.

Reply:
I could just duplicate these two catch statements, it's only 4 lines of code and the return code should ensure that it's not run twice.

Reply:
> I could just duplicate these two catch statements, it's only 4 lines of code and the return code should ensure that it's not run twice.

Hm... I'd be OK with that, but when we remove 2.7 & 3.5 support we have to keep that code and remove it further down. I'd probably settle on a TODO comment, just enough so we don't forget when we're done with the more sweeping changes.

For the returns, good to clean out later (maybe).

Reply:
@rwgk Happy to do that but it complicates the logic a bit.
I also realized I can't actually move it up or it will cause bugs if either of those errors are thrown in a nested exception (only 1 catch statement executes in C++ per throw to my knowledge). The other thing I wanted feedback is that I could disable this nested exception throwing with an optional potentially if there is interest.

Reply:
> I also realized I can't actually move it up or it will cause bugs if either of those errors are thrown in a nested exception (only 1 catch statement executes in C++ per throw to my knowledge).

oh ... now I'm thinking I need to read up on `std::nested_exception`.

big picture thinking: the simpler the better => I really hope we don't need an option to disable the feature.

Reply:
This is already opt-in, that is, if you don't use nested_exceptions in C++, you won't be using the feature here. (FWIU)

Reply:
Yeah refactored it now so the performance penalty is minimal (no extra std::rethrow)